source/validate_adjacency.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2017 LunarG Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Validates correctness of the intra-block preconditions of SPIR-V instructions.
+
+// TODO: Rename this file to validate_adjacency.cpp.
+
+#include "validate.h"
+
+#include <string>
+
+#include "diagnostic.h"
+#include "opcode.h"
+#include "val/instruction.h"
+#include "val/validation_state.h"
+
+namespace libspirv {
+
+spv_result_t ValidateAdjacency(ValidationState_t& _) {
+  auto const instructions = _.ordered_instructions();
+  for (auto i = instructions.cbegin(); i != instructions.cend(); ++i) {
+    switch (i->opcode()) {
+      case SpvOpPhi:
+        switch (prev(i)->opcode()) {
```

Comment:
It's worth saying that the instruction layout checks have ensured that prev(i) exists and is an instruction. (At least I hope that's true.)
I'd like to see a test case prove the validator is resilient to the case.

Reply:
At the very least we should be between OpFunction and OpFunctionEnd, so both prev() and next() should exist. It does make the code look flimsy though.

Reply:
Agreed. Added guards.